### ISSUE TITLE
Clarify pitch semantics for 2D image from buffer

### ISSUE THREAD

**Issue by jlewis-austin** [KhronosGroup/OpenCL-Docs#674]
For 2D images created from buffers in `clCreateImage`, the following train of logic leads to the conclusion that the row pitch must be set to zero in all cases:

1.  CL_MEM_USE_HOST_PTR or CL_MEM_ALLOC_HOST_PTR or CL_MEM_COPY_HOST_PTR are not valid flags for images created from other memory objects
2. _host_ptr_ must be NULL if those flags are not present: 
**CL_INVALID_HOST_PTR if _host_ptr_ is NULL and CL_MEM_USE_HOST_PTR or CL_MEM_COPY_HOST_PTR are set in _flags_ or if _host_ptr_ is not NULL but CL_MEM_COPY_HOST_PTR or CL_MEM_USE_HOST_PTR are not set in _flags_.**
 Although the 2D image does inherit HOST_PTR flag settings from the buffer object, the language here seems to refer to the actual italicized argument _flags_, not the inherited flag values. 
3. _image_row_pitch_ must be set to 0 if _host_ptr_ is NULL

But the last sentence of the _image_row_pitch_ description conflicts with that conclusion by suggesting that the pitch can be specified:
**For a 2D image created from a buffer, the pitch specified (or computed if pitch specified is 0) must be a multiple of the maximum of the CL_DEVICE_IMAGE_PITCH_ALIGNMENT value for all devices in the context associated with the buffer specified by _mem_object_ that support images.**

Questions:
1. Does the last sentence in the _image_row_pitch_ description override item 3 above (ie, _image_row_pitch_ doesn't need to be 0 in the image from buffer case when the _mem_object_ buffer had a CL_MEM_*_HOST_PTR flag)?
2. How is the pitch computed when _image_row_pitch_ is 0 in this case? The sentence 
**If host_ptr is not NULL and image_row_pitch = 0, image_row_pitch is calculated as image_width × size of element in bytes.**
 doesn't apply since _host_ptr_ is NULL in this case. One possible reading is that the computation forces the row pitch to a multiple of the alignment, but that is imprecise and doesn't give the app a guarantee of what multiple would be chosen as the actual pitch.

**Comment by bashbaug** [comment]
Here is an [approximate link](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#cl_image_desc) to the spec text in question.

It seems that the description for _image_row_pitch_ wasn't properly updated after adding the "2D image from buffer" case.  Do we just need to rephrase this section so it is clear that the _image_row_pitch_ is relevant if either (a) _host_ptr_ is non-NULL or (b) the image is created from a buffer, meaning that _buffer_ or _mem_object_ is non-NULL?  If so, that seems fairly straightforward.

Regarding:

> How is the pitch computed when image_row_pitch is 0 in this case?

Couldn't it be calculated the same, meaning "_image_width_ x size of element in bytes"?

**Comment by jlewis-austin** [comment]
@bashbaug both of your ideas are consistent with my interpretation, but we've seen some apps that rely on a different behavior, ie for a non-power-of-2 image width, the app passes `image_row_pitch = 0` and relies on the driver to compute a pitch meeting its alignment requirements (apparently the app doesn't need to know the actual pitch value for this use case, and therefore doesn't want to bother with querying for the proper `CL_DEVICE_IMAGE_PITCH_ALIGNMENT`). How the app knows how large of a buffer to allocate without knowing the pitch is an open question. I'm assuming they really do know and just bake that into their allocation calculations.

Since passing 0 for the pitch "just works" with other vendor(s), we're getting pressure to implement the same.

**Comment by bashbaug** [comment]
> Since passing 0 for the pitch "just works" with other vendor(s), we're getting pressure to implement the same.

It would be helpful to know a little more detail why _image_row_pitch_ equal to zero doesn't "just work" in your case.

The biggest difference when creating a 2D image from a buffer versus, say, initializing the contents of a 2D image from data in host memory is that the _image_row_pitch_ needs to be a multiple of `CL_DEVICE_IMAGE_PITCH_ALIGNMENT`.  So, if two devices have a different `CL_DEVICE_IMAGE_PITCH_ALIGNMENT`, it's possible that an _image_row_pitch_ is valid for one device but not for another.  This is the case regardless whether the _image_row_pitch_ is specified directly as part of the call or is zero and hence computed from the image width and the size of the image element in bytes.  Is that what is happening, or is it somethign different?  Thanks!

**Comment by jlewis-austin** [comment]
Yes, our implementation computes the _image_row_pitch_ from the image width and element size, then returns `CL_INVALID_IMAGE_FORMAT_DESCRIPTOR` when it fails the alignment check (another question--why is that the error for a pitch issue?). Since the image width is not a power of 2, it should fail on any implementation using this interpretation, regardless of differences in `CL_DEVICE_IMAGE_PITCH_ALIGNMENT`.

**Comment by bashbaug** [comment]
I see, so the other implementation should be failing and it isn't (unless, I suppose, its `CL_DEVICE_IMAGE_PITCH_ALIGNMENT` is equal to `1`).

Some actions we can take from this issue are:

1. Clarify the _image_row_pitch_ bullet point so it's clearer that the pitch can be nonzero when the image is created from a buffer, in addition to when the _host_ptr_ is non-NULL.  I agree that the current description is confusing.
2. Possibly change the error code for an invalid _image_row_pitch_ to [`CL_INVALID_IMAGE_DESCRIPTOR`](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_INVALID_IMAGE_DESCRIPTOR) rather than [`CL_INVALID_IMAGE_FORMAT_DESCRIPTOR`](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_INVALID_IMAGE_FORMAT_DESCRIPTOR).  I think this is a bug (going all the way back to the OpenCL 1.2 extension spec!), since clearly the image descriptor is invalid, not the image **format** descriptor.
3. Possibly ensure this scenario is covered by a future error checking layer?
4. Possibly considere a negative CTS test to ensure implementations are properly considering this condition?

(1) and (2) are spec issues and would be fixed here.  (3) and (4) would need to be fixed elsewhere.

**Comment by jlewis-austin** [comment]
I agree (1) and (2) should be addressed and shouldn't be controversial.

The remaining doc issue though is the question of what the computation should be if _image_row_pitch_ is 0. I see a fairly recent change in ROCm that implements the pitch alignment computation instead of "_image_width_ x size of element in bytes" in this case, so it appears there are multiple implementations choosing that interpretation. If we want to clarify this and explicitly specify one or the other, it seems pragmatically easier to go with the aligned computation, just because it shouldn't break any apps that rely on the other interpretation, whereas going the reverse direction does break apps (as we've found out). 

If we were to go the aligned route, we would need to specify the computation explicitly, as in aligning to the next larger multiple of `CL_DEVICE_IMAGE_PITCH_ALIGNMENT` vs the current requirement for user-provided pitches, which allows any multiple of the maximum `CL_DEVICE_IMAGE_PITCH_ALIGNMENT` to be specified.

**Comment by bashbaug** [comment]
Proposed updated description for `image_row_pitch`:

> * `image_row_pitch` is the scan-line pitch in bytes. If _host_ptr_ is not NULL or the image is a 2D image created from a buffer, `image_row_pitch` may be zero, in which case the image row pitch is calculated as `image_width` × the size of an image element in bytes. The image row pitch must be ≥ `image_width` × the size of an image element in bytes and must be a multiple of the size of an image element in bytes. For a 2D image created from a buffer, the image row pitch must be a multiple of the maximum of the `CL_DEVICE_IMAGE_PITCH_ALIGNMENT` value for all devices in the context that support images. For all other cases the `image_row_pitch` must be zero.

Note that this is intended to describe the existing spec behavior more clearly and is not intended to change the image row pitch calculation so it takes into account the `CL_DEVICE_IMAGE_PITCH_ALIGNMENT`.  I think we should discuss this separately from the clarification.

**Comment by Oblomov** [comment]
I would clarify what is meant by “For all other cases”, because at first read, given its position after “For a 2D image created from a buffer”, it would seem that “all other cases” would include also 2D images not created from buffer, which isn't the case obviously. Maybe something like:

> `image_row_pitch` is the scan-line pitch in bytes. It must be zero if `host_ptr` is NULL and the image is not a 2D image created from a buffer. If `image_row_pitch` is set to zero and `host_ptr` is not NULL or the image is a 2D image created from a buffer, then the image row pitch is calculated as image_width × the size of an image element in bytes. The image row pitch must be ≥ image_width × the size of an image element in bytes and must be a multiple of the size of an image element in bytes. For a 2D image created from a buffer, the image row pitch must also be a multiple of the maximum of the CL_DEVICE_IMAGE_PITCH_ALIGNMENT value for all devices in the context that support images.

I wonder if it might be worth it to add a small table illustrating the various combinations.

**Comment by jlewis-austin** [comment]
> I think we should discuss this separately from the clarification.

That sounds reasonable, though I'm a bit concerned about formalizing this to clearly make existing implementations non-conformant without resolving the question. We're most likely going to be forced to implement the alignment behavior for compatibility so don't want that to become a roadblock to conformance submission.

**Comment by Oblomov** [comment]
In the 2D image from buffer case, have you verified that the non-power-of-2 width case with zero image pitch _actually_ works with other vendors? I've had bad experiences with some vendors even in their proprietary API, when the equivalent call would not produce an error, but the fetched data was completely bogus because the alignment constraints was not satisfied.

**Comment by bashbaug** [comment]
> Maybe something like ...

Yes, thanks, I agree your wording is better.

I'll create a PR with these changes and I'll try adding a small table also.

Thanks!

**Comment by jlewis-austin** [comment]
> In the 2D image from buffer case, have you verified that the non-power-of-2 width case with zero image pitch _actually_ works with other vendors?

Good point, we'll follow up on that. The feedback was based on existing apps, though, so if it was bogus it seems like it should have affected their app behavior in a noticeable way.

**Comment by kpet** [comment]
Note that the extension spec currently says:

>When the 2D image from buffer is created, the client must specify the width, height and image format (i.e. channel order and channel data type). If these are not specified, **clCreateImage** returns a `NULL` value with `errcode_ret`
set to `CL_INVALID_IMAGE_FORMAT_DESCRIPTOR`. The pitch can be optionally specified. If the pitch is not
specified, the pitch is computed as width × bytes per pixel based on the image format.
The pitch specified (or computed if pitch specified is 0) must be a multiple of the maximum of the
`CL_DEVICE_IMAGE_PITCH_ALIGNMENT_KHR` value for all devices in the context associated with the buffer
that support images.
